src/NuGet.Core/NuGet.Protocol/HttpSource/HttpSourceAuthenticationHandler.cs

```diff
@@ -29,32 +29,29 @@ public class HttpSourceAuthenticationHandler : DelegatingHandler
         private Dictionary<string, AmbientAuthenticationState> _authStates = new Dictionary<string, AmbientAuthenticationState>();
         private HttpSourceCredentials _credentials;
 
+        /// <summary>
+        /// Creates a http source authentication hnadler
+        /// </summary>
+        /// <param name="packageSource"></param>
+        /// <param name="clientHandler"></param>
+        /// <param name="credentialService"></param>
+        /// <param name="useDefaultNetworkCredentials"></param>
+        /// <remarks>The useDefaultNetworkCredentials value needs to be evaluated by calling the CredentialService.HandlesDefaultCredentials() method. For performance considerations tha tmethod was made async, and can't be executed in the constructor</remarks>
         public HttpSourceAuthenticationHandler(
             PackageSource packageSource,
             HttpClientHandler clientHandler,
-            ICredentialService credentialService)
+            ICredentialService credentialService,
+            bool useDefaultNetworkCredentials)
             : base(clientHandler)
         {
-            if (packageSource == null)
-            {
-                throw new ArgumentNullException(nameof(packageSource));
-            }
-
-            _packageSource = packageSource;
-
-            if (clientHandler == null)
-            {
-                throw new ArgumentNullException(nameof(clientHandler));
-            }
-
-            _clientHandler = clientHandler;
+            _packageSource = packageSource ?? throw new ArgumentNullException(nameof(packageSource));
+            _clientHandler = clientHandler ?? throw new ArgumentNullException(nameof(clientHandler));
 
             // credential service is optional as credentials may be attached to a package source
             _credentialService = credentialService;
 
             // Create a new wrapper for ICredentials that can be modified
-
-            if (_credentialService == null || !_credentialService.HandlesDefaultCredentials)
+            if (useDefaultNetworkCredentials)
```

Comment:
I'm not sure I follow?

The ICredentials value needs to be set at construction time of the auth handler, because otherwise we'd have synchronization concerns. 

The methods that gets us the bool value is async and can't be called in a constructor.

Reply:
I see. You can still do Method.Wait() or Method.Result. But this is fine.

Reply:
You can't use method wait in constructors. 
Result could lead to deadlocks.